contracts/src/assertionStakingPool/AssertionStakingPool.sol

```diff
@@ -0,0 +1,147 @@
+
+    /// @param _rollup Rollup contract of target chain
+    /// @param _assertionInputs Inputs to be passed into Rollup.stakeOnNewAssertion
+    /// @param _assertionHash Assertion hash to be passed into Rollup.stakeOnNewAssertion
+    constructor(
+        address _rollup,
+        AssertionInputs memory _assertionInputs,
+        bytes32 _assertionHash
+    ) {
+        rollup = _rollup;
+        assertionHash = _assertionHash;
+        assertionInputs = _assertionInputs;
+        stakeToken = IERC20(IRollupCore(rollup).stakeToken());
+    }
+
+    /// @notice Deposit stake into pool contract. Callable only if target stake hasn't been reached and assertion has not been asserted yet.
+    /// @param _amount amount of stake token to deposit
+    function depositIntoPool(uint256 _amount) external {
+        if (poolState != PoolState.PENDING) {
+            revert PoolNotInPendingState(poolState);
+        }
+        uint256 requiredStake = getRequiredStake();
+
+        uint256 currentPoolBalance = stakeToken.balanceOf(address(this));
+        if (currentPoolBalance >= requiredStake) {
+            revert PoolStakeAlreadyReached(requiredStake);
+        }
+
+        // don't send more than necessary
+        uint256 amountToTransfer;
+        if (currentPoolBalance + _amount > requiredStake) {
+            amountToTransfer = requiredStake - currentPoolBalance;
+        } else {
+            amountToTransfer = _amount;
+        }
+
+        depositedTokenBalances[msg.sender] += amountToTransfer;
+        stakeToken.safeTransferFrom(msg.sender, address(this), amountToTransfer);
+        emit StakeDeposited(msg.sender, amountToTransfer);
+    }
+
+    /// @notice Create assertion. Callable only if required stake has been reached and assertion has not been asserted yet.
+    function createAssertion() external {
+        if (poolState != PoolState.PENDING) {
+            revert PoolNotInPendingState(poolState);
+        }
+
+        uint256 balance = stakeToken.balanceOf(address(this));
+        uint256 requiredStake = getRequiredStake();
+        if (balance < requiredStake) {
+            revert NotEnoughStake(balance, requiredStake);
+        }
+
+        poolState = PoolState.ASSERTED;
+        // approve spending from rollup for newStakeOnNewAssertion call
+        stakeToken.approve(rollup, requiredStake);
+        IRollupUser(rollup).newStakeOnNewAssertion(requiredStake, assertionInputs, assertionHash);
+        emit AssertionCreated();
+    }
+
+    /// @notice Move stake back from rollup contract to this contract. Calalble only if this contract has already created an assertion and it's been confirmed.
+    function returnOldStakeBackToPool() external {
+        if (poolState != PoolState.ASSERTED) {
+            revert PoolNotInAssertedState(poolState);
+        }
+
+        poolState = PoolState.CONFIRMED;
+
+        if (IRollupCore(rollup).getAssertion(assertionHash).status != AssertionStatus.Confirmed) {
```

Comment:
should allow refund as soon as the stake became inactive (have child or confirmed)

Reply:
it seems like that would create a griefing vector; i.e., `returnOldStakeBackToPool` has to be permissionless, which means griefer can make a child on pool's assertion and kick out the pool's stake.

Reply:
They can, but there is no harm. The assertion has already been made.

Reply:
i see